### DISCUSSION TITLE
Step operator

### DISCUSSION THREAD

**Discussion by AKHCE** [MilesCranmer/PySR#246] [Q&A] [answered]
Hi,
I want to define customized operators which have a "if-then-else" structure. For example, 
lambda x: 1 if x>0 else -1 if x<0 else 0
and
lambda x: 1 if x>0 else 0  if x<0 else 0.5

To do so in PySR, I wrote the following code:
```py
model = PySRRegressor(
    model_selection="best", 
    binary_operators=["*", "+", "-", "/"],
    unary_operators=["Mystep", "Mysign"],
    extra_sympy_mappings={"Mystep": lambda x: 1 if x>0 else 0  if x<0 else 0.5, 
                                            "Mysign": lambda x: 1 if x>0 else -1 if x<0 else 0},
)
```

But, it doesn't work!
Any advice is much appreciated

**Comment by MilesCranmer** [comment, marked as answer]
Here's an implementation which might be what you're looking for: https://github.com/MilesCranmer/PySR/discussions/125

Essentially you have to use `sympy.Piecewise`, rather than Python's `if`, in order to get a correct mapping to `sympy`

**Reply by MilesCranmer** [reply]
There's also `sympy.Heaviside`

**Reply by AKHCE** [reply]
Thank you so much for your prompt response.
Would you please do a favor a give an example on how we can add a Heaviside function as an operator into PySRRegressor (...)? I have difficulty in implementation !
I appreciate your answer much

**Reply by MilesCranmer** [reply]
Does this work?
```
unary_operators=["heaviside(x::T) where {T} = T(x>=0)"],
extra_sympy_functions={"heaviside": sympy.Heaviside}
```

**Reply by AKHCE** [reply]
I do really appreciate your speedy reply. Yes, it is working, and I see heaviside(.) in correlation. Just wondering if you could please give a bit more elaboration on "T(x>=0)", what does this say? I mean what if x<0?
Many thanks

**Reply by MilesCranmer** [reply]
That is Julia syntax for a function. See https://docs.julialang.org/en/v1/manual/methods/#Parametric-Methods for more details.

`x>=0` will turn into `true` if true, `false` if false.

Then, the `T` is whatever type the input is. WIth default settings in PySR, `T` would become a `Float32`. 

Wrapping a value with a type, like `Float32(x)` would perform a conversion. Thus, `Float32(true)` would become `1.0f0`, and `Float32(false)` would become `0.0f0`.

**Reply by AKHCE** [reply]
I got it!
Thanks

**Reply by MilesCranmer** [reply]
Also, an alternate and maybe faster version would be:

```julia
heaviside(x::T) where {T} = x>=0 ? one(T) : zero(T)
```

By the way, if you are up for it, I'd welcome a PR adding this as a built-in operator!  (so people don't have to define it themselves.)

**Reply by AKHCE** [reply]
Yes, please.
Your library is literally AWESOME ! and salient point is that I as a general user don't need to dig into detailed programming. So, then I recommend enriching operators a bit further e.g. heaviside.
Thank you so much
